### PR TITLE
Add ability to get symbolic dimension info for graph inputs and outputs.

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Add a function to be able to get the symbolic dimension names. 

**Motivation and Context**
All symbolic dimensions currently use -1 in TensorShape, which doesn't provide a way to know which specific symbolic dimension the -1 represents. Adding a method to return the dim_param value for each dimension allows a user to access this information. This is also more consistent with how the python API works.

### CODE REVIEW

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 1]
```diff
@@ -446,6 +446,7 @@ struct OrtApi {
   OrtStatus*(ORT_API_CALL* GetTensorElementType)(_In_ const OrtTensorTypeAndShapeInfo*, _Out_ enum ONNXTensorElementDataType* out)NO_EXCEPTION;
   OrtStatus*(ORT_API_CALL* GetDimensionsCount)(_In_ const OrtTensorTypeAndShapeInfo* info, _Out_ size_t* out)NO_EXCEPTION;
   OrtStatus*(ORT_API_CALL* GetDimensions)(_In_ const OrtTensorTypeAndShapeInfo* info, _Out_ int64_t* dim_values, size_t dim_values_length)NO_EXCEPTION;
+  OrtStatus*(ORT_API_CALL* GetSymbolicDimensions)(_In_ const OrtTensorTypeAndShapeInfo* info, _Out_ const char** dim_params, size_t dim_values_length)NO_EXCEPTION;
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h]:
```suggestion
  OrtStatus*(ORT_API_CALL* GetSymbolicDimensions)(_In_ const OrtTensorTypeAndShapeInfo* info, _Out_ const char** dim_params, size_t dim_params_length)NO_EXCEPTION;
```

File: onnxruntime/core/framework/tensor_type_and_shape.h [hunk 2]
```diff
@@ -6,6 +6,9 @@ struct OrtTensorTypeAndShapeInfo {
  public:
   ONNXTensorElementDataType type = ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT;
   onnxruntime::TensorShape shape;
+  // dim_param values. empty string if dim_value or no dim_param was specified.
+  // one entry per dimension in shape. only guaranteed to be populated for graph inputs and outputs
+  std::vector<std::string> dim_params;
```
Comment (pranavsharma) [inline onnxruntime/core/framework/tensor_type_and_shape.h]: Should this be part of TensorShape instead so that it can be used in other places as well?

**Comment by pranavsharma** [comment]
We should probably add an API to return denotations as well. This could be useful when calling OrtAddFreeDimensionOverride which works on denotations.

File: onnxruntime/core/framework/tensor_type_and_shape.h (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/framework/tensor_type_and_shape.h]: I considered it but I don't think so. Adding copies of strings to all TensorShape instances has a good chance of causing performance issues. We have seen perf improvements from simply avoiding the copy of the std::vector of dims in the ctor.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 1 above)
Comment (skottmckay) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Will update ort_apis.h as well.

**Comment by skottmckay** [comment]
> We should probably add an API to return denotations as well. This could be useful when calling OrtAddFreeDimensionOverride which works on denotations.

Agreed. Added Task 524263 to the backlog. 

Side question: Why does 'OrtAddFreeDimensionOverride' have 'Ort' as a prefix?

File: onnxruntime/core/framework/tensor_type_and_shape.h (hunk 2 above)
Comment (pranavsharma) [reply onnxruntime/core/framework/tensor_type_and_shape.h]: Ok. I was assuming that most of the time you would've symbolic dims in the input tensors only and this vector would be empty in all other cases.

**Comment by pranavsharma** [comment]
> > We should probably add an API to return denotations as well. This could be useful when calling OrtAddFreeDimensionOverride which works on denotations.
> 
> Agreed. Added Task 524263 to the backlog.
> 
> Side question: Why does 'OrtAddFreeDimensionOverride' have 'Ort' as a prefix?

Good point. Before Ryan did the refactoring to add versioning support all APIs had the Ort prefix. I believe this was missed as part of the refactoring. We should fix this.

**Comment by pranavsharma** [comment]
LGTM 👍

**Pull request merged**